src/Workspaces/Remote/ServiceHub/Services/BrokeredServiceBase.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.ServiceHub.Framework.Services;
+using Nerdbank.Streams;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Base type for Roslyn brokered services hosted in ServiceHub.
+    /// </summary>
+    internal abstract class BrokeredServiceBase : IDisposable
+    {
+        internal interface IFactory
+        {
+            public object Create(IDuplexPipe pipe, IServiceProvider hostProvidedServices, ServiceActivationOptions serviceActivationOptions, IServiceBroker serviceBroker);
+            public Type ServiceType { get; }
+        }
+
+        internal readonly struct ServiceConstructionArguments
+        {
+            public readonly IServiceProvider ServiceProvider;
+            public readonly IServiceBroker ServiceBroker;
+            public readonly CancellationTokenSource ClientDisconnectedSource;
+
+            public ServiceConstructionArguments(IServiceProvider serviceProvider, IServiceBroker serviceBroker, CancellationTokenSource clientDisconnectedSource)
+            {
+                ServiceProvider = serviceProvider;
+                ServiceBroker = serviceBroker;
+                ClientDisconnectedSource = clientDisconnectedSource;
+            }
+        }
+
+        internal abstract class FactoryBase<TService> : IServiceHubServiceFactory, IFactory
```

Comment:
very generic name.  can this be ServiceHubeServiceFactoryBase?

Reply:
Not necessary since it's already in `BrokeredServiceBase` type.